eparticlectr/eparticlectr.cpp

```diff
@@ -433,6 +447,154 @@ void eparticlectr::finalize( uint64_t proposal_id ) {
         proptable.erase( prop_it );
 }
 
+[[eosio::action]]
+void eparticlectr::finalizeextr( uint64_t proposal_id ) {
+    // Verify proposal exists
+    propstblex proptable( _self, _self.value );
+    auto prop_it = proptable.find( proposal_id );
+    eosio::check( prop_it != proptable.end(), "Proposal not found" );
+
+    // Verify voting period is complete
+    eosio::check( eosio::current_time_point().sec_since_epoch() > prop_it->endtime, "Voting period is not over yet");
+
+    // Retrieve votes from DB
+    votestblex votetbl( _self, proposal_id );
+    auto vote_it = votetbl.begin();
+    eosio::check( vote_it != votetbl.end(), "No votes found for proposal");
+
+    // Tally votes
+    uint64_t yes_votes = 0;
+    uint64_t no_votes = 0;
+    while(vote_it->proposal_id == proposal_id && vote_it != votetbl.end()) {
+        if (vote_it->approve)
+            yes_votes += vote_it->amount;
+        else
+            no_votes += vote_it->amount;
+        vote_it++;
+    }
+
+    // Determine slashing conditions
+    vote_it = votetbl.begin();
+    bool approved = 0;
+    bool istie = 0;
+    float totalVotes = yes_votes + no_votes;
+    if ((yes_votes / totalVotes) >= TIER_ONE_THRESHOLD){
+        approved = 1;
+        if ((yes_votes / totalVotes) == TIER_ONE_THRESHOLD){
+            istie = 1;
+        }
+    }
+    float slash_ratio = 0.0f;
+    if (approved){
+        slash_ratio = (yes_votes - no_votes) / totalVotes;
+    }
+    else{
+        slash_ratio = (no_votes - yes_votes) / totalVotes;
+    }
+
+    // Make sure no weird bugs cause the slash reward to under/overflow
+    eosio::check( slash_ratio >= 0.0f && slash_ratio <= 1.0f, "Slash ratio out of bounds");
+
+    rewardstblex rewardstable( _self, _self.value );
+    staketblex staketable(_self, _self.value);
+
+    // Slash / reward votes
+    // Tally vote points
+    uint64_t total_vote_points = 0;
+    while(vote_it != votetbl.end() && istie == 0) {
```

Comment:
here there is a second iteration over the vote table, again same issue that pointed before, there are chances to get timeouts in this action.

Reply:
will leave for now, but your point is valid